File: connectors/java/lom/LomODataClient.java
// SPDX-License-Identifier: Apache-2.0
package lom;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;

/** Thin HTTP client for lomd (OpenAPI / OData). */
public final class LomODataClient implements AutoCloseable {
	private final HttpClient http = HttpClient.newHttpClient();
	private final String base;
	private final String apiKey;

	public LomODataClient(String baseUrl, String apiKey) {
		this.base = baseUrl.replaceAll("/$", "");
		this.apiKey = apiKey;
	}

	private String get(String path) throws Exception {
		HttpRequest req = HttpRequest.newBuilder(URI.create(base + "/" + path))
			.header("X-Api-Key", apiKey)
			.GET()
			.build();
		HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString());
		if(res.statusCode() >= 300) {
			throw new IllegalStateException("HTTP " + res.statusCode() + ": " + res.body());
		}
		return res.body();
	}

	public String health() throws Exception {
		return get("health");
	}

	public String listPeople(String filter, int top) throws Exception {
		String q = "api/ListPeople?$top=" + top;
		if(filter != null && !filter.isEmpty()) {
			q += "&$filter=" + URLEncoder.encode(filter, StandardCharsets.UTF_8);
		}
		return get(q);
	}

	private String send(String method, String path, String body) throws Exception {
		HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(base + "/" + path))
			.header("X-Api-Key", apiKey)
			.header("Content-Type", "application/json");
		if(body != null) b.method(method, HttpRequest.BodyPublishers.ofString(body));
		else b.method(method, HttpRequest.BodyPublishers.noBody());
		HttpResponse<String> res = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
		if(res.statusCode() >= 300) {
			throw new IllegalStateException("HTTP " + res.statusCode() + ": " + res.body());
		}
		return res.body();
	}

	public String query(String selector) throws Exception {
		return send("POST", "lom/query", "{\"selector\":\"" + selector + "\"}");
	}

	public String createPerson(String jsonBody) throws Exception {
		return send("POST", "api/CreatePeople", jsonBody);
	}

	public String updatePerson(String id, String jsonBody) throws Exception {
		return send("PATCH", "odata/People('" + id + "')", jsonBody);
	}

	public String deletePerson(String id) throws Exception {
		return send("DELETE", "odata/People('" + id + "')", null);
	}

	@Override
	public void close() {}
}

File: connectors/csharp/Lom.OData/LomODataClient.cs
// SPDX-License-Identifier: Apache-2.0
using System;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace Lom.OData
{
	/// <summary>HTTP client for lomd (OpenAPI / OData surface).</summary>
	public sealed class LomODataClient : IDisposable
	{
		readonly HttpClient _http;

		public LomODataClient(string baseUrl, string apiKey)
		{
			_http = new HttpClient { BaseAddress = new Uri(baseUrl.TrimEnd('/') + "/") };
			_http.DefaultRequestHeaders.Add("X-Api-Key", apiKey);
		}

		public async Task<string> HealthAsync()
		{
			return await _http.GetStringAsync("health");
		}

		public async Task<string> ListPeopleAsync(string? filter = null, int top = 50)
		{
			var q = "api/ListPeople?$top=" + top;
			if(!string.IsNullOrEmpty(filter))
				q += "&$filter=" + Uri.EscapeDataString(filter);
			return await _http.GetStringAsync(q);
		}

		public async Task<string> ODataPeopleAsync(string? filter = null, int top = 50)
		{
			var q = "odata/People?$top=" + top;
			if(!string.IsNullOrEmpty(filter))
				q += "&$filter=" + Uri.EscapeDataString(filter);
			return await _http.GetStringAsync(q);
		}

		public async Task<string> QueryAsync(string? selector = null, string? xpath = null)
		{
			var body = "{";
			if(!string.IsNullOrEmpty(selector)) body += "\"selector\":\"" + selector + "\"";
			else if(!string.IsNullOrEmpty(xpath)) body += "\"xpath\":\"" + xpath + "\"";
			body += "}";
			var resp = await _http.PostAsync("lom/query", new StringContent(body, Encoding.UTF8, "application/json"));
			return await resp.Content.ReadAsStringAsync();
		}

		public async Task<string> CreatePersonAsync(string jsonBody)
		{
			var resp = await _http.PostAsync("api/CreatePeople", new StringContent(jsonBody, System.Text.Encoding.UTF8, "application/json"));
			return await resp.Content.ReadAsStringAsync();
		}

		public async Task<string> UpdatePersonAsync(string id, string jsonBody)
		{
			var req = new HttpRequestMessage(HttpMethod.Patch, "odata/People('" + id + "')")
			{
				Content = new StringContent(jsonBody, System.Text.Encoding.UTF8, "application/json")
			};
			var resp = await _http.SendAsync(req);
			return await resp.Content.ReadAsStringAsync();
		}

		public async Task DeletePersonAsync(string id)
		{
			var resp = await _http.DeleteAsync("odata/People('" + id + "')");
			resp.EnsureSuccessStatusCode();
		}

		public void Dispose() => _http.Dispose();
	}
}
